### DISCUSSION TITLE
Supported Kura 5.5.0 profiles and future plans

### DISCUSSION THREAD

**Discussion by pierantoniomerlino** [eclipse-kura/kura#5224] [Ideas]
Hello everyone!

This post is to share with everyone the plans for the supported profiles and devices in the upcoming Kura 5.5.

With the introduction of the **generic** profiles in Kura 5.3, we were able to decouple a profile from the hardware that runs Kura. So, the only thing to do is to pick up the Kura installer for the correct architecture (arm32/64 or x86_64) and install it in your device. The idea was to slowly drop the profiles specific to a particular board and release only installers tailored for a specific hardware architectures.

Currently (Kura **5.4**) the supported profiles/devices are the following:

Device |	Architecture |	OS |
-------|---------------|--------------------------|
Raspberry Pi 3/4 |	arm32 |	Raspbian "Bullseye" |
Raspberry Pi 3/4 |	arm64 |	Ubuntu 20.04 |
Intel Up² |	x86_64 |	Ubuntu 20.04 |
NVIDIA Jetson Nano™ | arm64 |	Ubuntu 18.04 |
generic-arm32 | arm32 | generic Debian OS |
generic-arm64 | arm64 | generic Debian OS |
generic-x86_64 | x86_64 | generic Debian OS |

The upcoming Kura **5.5** will drop the Rasperry Pi profiles, replacing them with the generic versions. The Intel Up² and NVIDIA Jetson Nano™ ones will be deprecated and removed in Kura 5.6. So, the resulting table for Kura **5.5** will be:

Device |	Architecture |	OS |
-------|---------------|--------------------------|
Intel Up² |	x86_64 |	Ubuntu 20.04 |
NVIDIA Jetson Nano™ | arm64 |	Ubuntu 18.04 |
generic-arm32 | arm32 | generic Debian OS |
generic-arm64 | arm64 | generic Debian OS |
generic-x86_64 | x86_64 | generic Debian OS |

**Comment by alexcorvis84** [comment]
Hi @pierantoniomerlino, just a question.

Regarding the profiles for RaspberryPi, as far as I know they included support for the GPIO (_jdk.dio.properties_ specific file and related).

So, this new approach moving for generic profiles means that there won't be any support for Kura to manage RPi GPIOs?

Thanks in advance!

**Reply by pierantoniomerlino** [reply]
The GPIO will be supported for RPi, UP2 and Nano. There is a custom script in the generic profiles installer that copies the correct `jdk.dio.properties` if a known device is detected.

However, even in a unknown device, the user can manually fill that file and the GPIOs will be correctly detected by Kura.

**Reply by alexcorvis84** [reply]
Thanks for the information @pierantoniomerlino !

**Reply by alexcorvis84** [reply]
Hi @pierantoniomerlino, I think there's a typo error in the tables for the generic-arm64 Architecture. I guess they should be arm64 for both (instead of arm32), right?

**Comment by alexcorvis84** [comment]
Hi @pierantoniomerlino, 

I have checked that in Kura 5.4.0, the installation of Java takes as preference Java17 instead of Java8.

As far as I've checked, the **control** file from the **.deb** package, contains the following:

```
Package: kura
Version: 5.4.0
Section: misc
Priority: low
Architecture: all
Depends: openjdk-8-jre-headless | temurin-8-jdk | openjdk-17-jre-headless | temurin-17-jdk,
 setserial, zip, gzip, unzip, procps, usbutils, socat, gawk, sed, inetutils-telnet,
 polkit | policykit-1, ssh | openssh, openssl, busybox, openvpn,
 bluez | bluez5, bluez-hcidump | bluez5-noinst-tools,
 ntpdate, chrony, chronyc | chrony, cron | cronie,
 network-manager | networkmanager, bind9 | bind, dnsmasq | isc-dhcp-server | dhcp-server, dnsmasq | isc-dhcp-server | dhcp-client, iw, iptables, modemmanager,
 logrotate,
 gpsd
Recommends: python3
Installed-Size: 98509
Maintainer: support@eurotech.com
Description: Kura is an inclusive software framework that puts a layer
 between the operating system and the customer application, with industry
 standard interfaces that shorten custom development time, simplified coding
 and software that can be easily ported from one hardware platform
 to another.
```

As I understand it, the _**Depends**_ field requires one of the dependencies to be installed (trying in order). However, installing the package on a RaspberryPi 4 with Raspberry Pi OS Lite 32-bit, Debian Bookworm, it installs Java17 instead of Java8.

![image](https://github.com/eclipse/kura/assets/44984326/c7a8d935-f074-4412-a4e6-c672579f3bf1)


Why does it install Java17 and not Java8, where does it take that preference then?

Thanks in advance.

**Reply by pierantoniomerlino** [reply]
Interesting... 

In this [documentation](https://www.debian.org/doc/manuals/aptitude/ch02s03s02.en.html) at the section 2, the rule is:

>  If the dependency is on several packages combined with OR, examine each of the alternatives in the order that they are given. For instance, if a package depends on “exim | mail-transport-agent”, aptitude will first process exim, then mail-transport-agent. 

So, I expect that Java8 is preferred to Java17. Is Java8 available on that OS?

**Reply by alexcorvis84** [reply]
Yeah, exactly, I looked into the same link and that's why I expected that Java8 is preferred to Java17, but as you can see it's not... 

> Is Java8 available on that OS?

The thing is that I'm installing it in a fresh SD card (created with Official Raspberry Pi Imager software, Raspbian OS Bookworm 32-bit or 64-bit Lite versions) which they come without any java sdk preinstalled.

PD: Btw, I can install later manually Java8 and select as it as manual preferred (sudo update-alternatives --config java), then restart Kura and run it with Java8... but, as you can see, the **default** installation opts for Java17 instead (which _seems_ both did not expected! 😅)

I'm just trying to figure it out/understand what's the reason behind this behaviour...

Many thanks for your quick feedback.

**Reply by alexcorvis84** [reply]
Hi @pierantoniomerlino 

I can confirm that the 'issue' seems related to **Raspberry Pi OS version**....

I've just tried to install the generic arm32 version in a Raspberry Pi OS **(Bullseye)** 32-bit Lite and I can confirm it defaults to install **Java8** instead of Java17. 

So something weird is happening on latest **(Bookworm)** version of Raspberry Pi OS...

![image](https://github.com/eclipse/kura/assets/44984326/ccdc41ee-4f4e-42c5-88ca-c3030d229618)

**Comment by alexcorvis84** [comment]
Hi @pierantoniomerlino ,

I've checked that Intel Up² and NVIDIA Jetson Nano™ profiles are [still present](https://download.eclipse.org/kura/releases/5.6.0/) in **5.6.0**... So, i guess they will be deprecated in the next 5.6.X version?

Many thanks in advance and happy X-mas!

**Reply by mattdibi** [reply]
Hello there @alexcorvis84,

Intel Up² and NVIDIA Jetson Nano™ profiles will be removed in 6.0.0 (which is our next planned release) along with the arm32 profile. You can take a look at our recent PRs about this:
- https://github.com/eclipse-kura/kura/pull/5615
- https://github.com/eclipse-kura/kura/pull/5315

An overall picture of the planned changes is available here: https://github.com/eclipse-kura/kura/discussions/5410

Hope this helps and happy xmas to you too :)

**Reply by alexcorvis84** [reply]
Thank you @mattdibi!
